File: storekit-offering-completing-and-restoring-in-app-purchases/Objective-C/macOS/PurchasesDetails.m
/*
See the LICENSE.txt file for this sample’s licensing information.

Abstract:
The table view controller for presenting details about a purchase.
*/

@import StoreKit;
#import "StoreManager.h"
#import "PurchasesDetails.h"
#import "NSDateFormatter+DateFormatter.h"

@interface PurchasesDetails ()
@property (weak) IBOutlet NSStackView *stackView;
@property (weak) IBOutlet NSTextField *productID;
@property (weak) IBOutlet NSTextField *transactionID;
@property (weak) IBOutlet NSTextField *transactionDate;

@property (weak) IBOutlet NSBox *originalTransaction;
@property (weak) IBOutlet NSTextField *originalTransactionID;
@property (weak) IBOutlet NSTextField *originalTransactionDate;
@end

@implementation PurchasesDetails
#pragma mark - View Life Cycle

- (void)viewDidAppear {
    [super viewDidAppear];
    
    [self hideBox:self.originalTransaction forStatus:YES];
    [self reloadTableAndSelectFirstRowIfNecessary];
}

#pragma mark - Update UI

/// Hides an NSBox object if the specified status is true. Shows it, otherwise.
-(void)hideBox:(NSBox *)box forStatus:(BOOL)status {
    if (self.viewLoaded) {
        box.hidden = status;
        [self.view layoutSubtreeIfNeeded];
    }
}

/// Reloads the UI with new products, invalid identifiers, and purchases.
-(void)reloadUIWithData:(NSMutableArray *)data {
    self.data = data;
    [self.tableView reloadData];
    [self reloadTableAndSelectFirstRowIfNecessary];
}

/// Reloads the table view and programmatically selects a purchase.
-(void)reloadTableAndSelectFirstRowIfNecessary {
    // Select the first purchase and display its information if no row is currently in a selected state. Display the current selection, otherwise.
    NSIndexSet *selection = (self.tableView.selectedRowIndexes.count == 0) ? [NSIndexSet indexSetWithIndex:0] : self.tableView.selectedRowIndexes;
    
    [self.tableView reloadData];
    [self.tableView selectRowIndexes:selection byExtendingSelection:NO];
}

#pragma mark - NSTableViewDelegate

- (NSView *)tableView:(NSTableView *)tableView viewForTableColumn:(NSTableColumn *)tableColumn row:(NSInteger)row {
    SKPaymentTransaction *transaction = (SKPaymentTransaction *)self.data[row];
    NSTableCellView *cell = [tableView makeViewWithIdentifier:tableColumn.identifier owner:self];
    
    if (cell != nil) {
        // Display the product's title associated with the payment's product identifier.
        cell.textField.stringValue = [[StoreManager sharedInstance] titleMatchingPaymentTransaction:transaction];
        return cell;
    }
    return nil;
}

/// Displays information about the selected purchase or a restored one.
- (void)tableViewSelectionDidChange:(NSNotification *)notification {
    NSInteger selectedRow = (self.tableView).selectedRow;
    NSDateFormatter *dateFormatter = [NSDateFormatter longStyle];
    
    [self hideBox:self.originalTransaction forStatus:YES];
    
    if (selectedRow >= 0 && (self.data.count > 0)) {
        SKPaymentTransaction *transaction = (SKPaymentTransaction *)self.data[selectedRow];
        
        self.productID.stringValue = transaction.payment.productIdentifier;
        self.transactionID.stringValue = transaction.transactionIdentifier;
        self.transactionDate.stringValue = [dateFormatter stringFromDate:transaction.transactionDate];
                
        // Display restored transactions if they exist.
        if (transaction.originalTransaction != nil) {
            [self hideBox:self.originalTransaction forStatus:NO];
            
            self.originalTransactionID.stringValue = transaction.originalTransaction.transactionIdentifier;
            self.originalTransactionDate.stringValue = [dateFormatter stringFromDate:transaction.originalTransaction.transactionDate];
        }
    }
}

@end

File: storekit-offering-completing-and-restoring-in-app-purchases/Swift/macOS/PurchasesDetails.swift
/*
See the LICENSE.txt file for this sample’s licensing information.

Abstract:
The table view controller for presenting details about a purchase.
*/

import Cocoa
import StoreKit

class PurchasesDetails: PrimaryViewController {
    // MARK: - Properties
    
    @IBOutlet weak fileprivate var stackView: NSStackView!
    @IBOutlet weak fileprivate var productID: NSTextField!
    @IBOutlet weak fileprivate var transactionID: NSTextField!
    @IBOutlet weak fileprivate var transactionDate: NSTextField!
        
    @IBOutlet weak fileprivate var originalTransaction: NSBox!
    @IBOutlet weak fileprivate var originalTransactionID: NSTextField!
    @IBOutlet weak fileprivate var originalTransactionDate: NSTextField!
    
    // MARK: - View Life Cycle
    
    override func viewDidAppear() {
        super.viewDidAppear()
        originalTransaction.hide()
        reloadTableAndSelectFirstRowIfNecessary()
    }
    
    // MARK: - Update UI
    
    /// Refreshes the UI with new payment transactions.
    func reload(with transactions: [SKPaymentTransaction]) {
        data = transactions
        self.tableView.reloadData()
        reloadTableAndSelectFirstRowIfNecessary()
    }
    
    /// Reloads the table view and programmatically selects a purchase.
    fileprivate func reloadTableAndSelectFirstRowIfNecessary() {
        // Select the first purchase and display its information if no row is currently in a selected state. Display the current selection, otherwise.
        let selection = (tableView.selectedRowIndexes.isEmpty) ? IndexSet(integer: 0) : tableView.selectedRowIndexes
        
        tableView.reloadData()
        tableView.selectRowIndexes(selection, byExtendingSelection: false)
    }
    
    // MARK: - NSTableViewDelegate
    
    func tableView(_ tableView: NSTableView, viewFor tableColumn: NSTableColumn?, row: Int) -> NSView? {
        guard let tableColumn = tableColumn, let cell = tableView.makeView(withIdentifier: tableColumn.identifier, owner: nil) as? NSTableCellView,
            let transaction = data[row] as? SKPaymentTransaction else { return nil }
        
        // Display the product's title associated with the payment's product identifier.
        cell.textField?.stringValue = StoreManager.shared.title(matchingPaymentTransaction: transaction)
        return cell
    }
    
    /// Displays information about the selected purchase or a restored one.
    func tableViewSelectionDidChange(_ notification: Notification) {
        let selectedRow = tableView.selectedRow
        guard (selectedRow >= 0) && (!data.isEmpty) else { return }
        
        originalTransaction.hide()
        
        guard let transaction = data[selectedRow] as? SKPaymentTransaction else { return }
        
        productID.stringValue = transaction.payment.productIdentifier
        transactionID.stringValue = transaction.transactionIdentifier!
        transactionDate.stringValue = DateFormatter.long(transaction.transactionDate!)
        
        // Display restored transactions if they exist.
        guard let transactionIdentifier = transaction.original?.transactionIdentifier, let transactionDate = transaction.original?.transactionDate
            else { return }
        
        originalTransaction.show()
        originalTransactionID.stringValue = transactionIdentifier
        originalTransactionDate.stringValue = DateFormatter.long(transactionDate)
    }
}

